internal/revision/scanner.go

```diff
@@ -0,0 +1,124 @@
+package revision
+
+import (
+	"bufio"
+	"io"
+	"unicode"
+)
+
+var zeroRune = rune(0)
+
+// scanner represents a lexical scanner.
+type scanner struct {
+	r *bufio.Reader
```

Comment:
I think using a bufio.Reader is the wrong approach here.

The only functionality your are using from a bufio.Reader is the `readRune` and `unreadRune` so using a full bufio.Reader cannot be justified.

Why are you not using a `text.Scanner` instead? (The peek and error reporting seems very helpful to achieve your goals here).